lib/active_merchant/billing/gateways/redsys.rb

```diff
@@ -333,19 +333,21 @@ def add_external_mpi_fields(data, options)
         return unless options[:three_d_secure]
 
         if options[:three_d_secure][:version] == THREE_DS_V2
-          data[:threeDSServerTransID] = options[:three_d_secure][:xid] if options[:three_d_secure][:xid]
+          data[:threeDSServerTransID] = options[:three_d_secure][:three_ds_server_trans_id] if options[:three_d_secure][:three_ds_server_trans_id]
+          unless data[:threeDSServerTransID]
```

Comment:
🤔 I'm not sure that we want to default this back to XID since that is specifically a field meant for 3DS V1. Is this primarily for backwards compatibility?

Reply:
I am not sure if it is needed. I just didn't want to remove it in case somebody already uses this gateway and passes xid. Do you think it is safe to remove xid from 3ds V2?

Reply:
Agree with Tatsiana, there is no reason to not leave it for backward compatibility.

Reply:
I might suggest this more explicit structure for reasonability, since the situations are mutually exclusive:
```
if options[:three_d_secure][:three_ds_server_trans_id]
  data[:threeDSServerTransID] = options[:three_d_secure][:three_ds_server_trans_id]
else
   data[:threeDSServerTransID] = options[:three_d_secure][:xid] if options[:three_d_secure][:xid]
end
```

Reply:
Alternately, I think it could be safely condensed into
`data[:threeDSServerTransID] = options.dig(:three_d_secure, :three_ds_server_trans_id) || options.dig(:three_d_secure, :xid)`

Reply:
I looked more into this. I'll remove assignment of xid to threeDSServerTransID because it doesn't look correct. Firstly, as you mentioned xid is for version 1. Secondly, xid and threeDSServerTransID are different identifiers. xid is for Directory Server, threeDSServerTransID is for 3ds Server.

Reply:
Void the comments above, I was thinking this was for the `dsTransID` field.

Reply:
Thanks for suggestion anyway. It will be useful for future because I wasn't sure about more concise way I can do it in Ruby.

Reply:
`dig` is somewhat new in Ruby. It's a more convenient way to safely reach into a nested hash, so it can replace instances of "options[:field] if options[:field]" etc., but since it is new it is not yet used extensively in AM.